### PR TITLE
[v2] compile cache dir

### LINKED ISSUE #4402: [v2] Cache Dir should probably be compiled down somewhat :/
There are some flow types in `cache-dir` babel@6 used to take care of them i think with the react-preset, v7 doesn't. We don't compile `cache-dir` having relied on letting the user's site handle it. We probably should do a pass over it to at least remove non-standard features, as well as outline in the Docs what the minimal babel setup should be (I think Kyle is already on that one)

### PR DESCRIPTION
**Pull request by pieh**
This moves `cache-dir` to `src` and runs it through babel as part of build / packaging process. After building dir structure is same as before, so there are no problems with `gatsby-dev-cli`

Closes #4402

### CODE REVIEW

**Comment by pieh** [comment]
TBH I'm not sure if this is what You had in mind in that ticket - but this seems reasonable?

File: packages/gatsby/src/cache-dir/static-entry.js [hunk 1]
```diff
@@ -195,11 +188,6 @@ export default (pagePath, callback) => {
     pathPrefix,
   })
 
-  const runtimeRaw = fs.readFileSync(
-    join(process.cwd(), `public`, runtimeScript),
-    `utf-8`
-  )
-
```
Comment (pieh) [inline packages/gatsby/src/cache-dir/static-entry.js]: this isn't currently used anywhere - should it be used?

File: packages/gatsby/src/cache-dir/json-store.js [hunk 2]
```diff
@@ -47,10 +42,9 @@ class JSONStore extends React.Component {
     }
   }
 
-  componentWillReceiveProps(nextProps) {
+  componentDidUpdate() {
```
Comment (pieh) [inline packages/gatsby/src/cache-dir/json-store.js]: not sure about that change, I've used recomendation from https://reactjs.org/docs/react-component.html#unsafe_componentwillreceiveprops , but it probably means extra render

**Comment by m-allanson** [comment]
Trying this with gatsbygram I get the following error from `yarn run develop`:

```
success createPages — 0.192 s
error UNHANDLED REJECTION


  Error: ENOENT: no such file or directory, lstat '/Users/mike/dev/gatsby-review/examples/gats  bygram/node_modules/gatsby/dist/internal-plugins/dev-404-page/raw_dev-404-page.js'

error Command failed with exit code 1.
```

File: packages/gatsby/package.json [hunk 3]
```diff
@@ -173,10 +173,11 @@
     "graphql": "^0.13.2"
   },
   "scripts": {
-    "build": "rimraf dist && npm run build:src && npm run build:internal-plugins && npm run build:rawfiles",
+    "build": "rimraf dist && rimraf cache-dir && npm run build:src && npm run build:internal-plugins && npm run build:rawfiles && build:cache-dir",
```
Comment (m-allanson) [inline packages/gatsby/package.json]: oh - this is missing an `npm run` at the end:  `&& build:cache-dir`

Comment (m-allanson) [reply packages/gatsby/package.json]: urgh I was looking at an old commit 🤦‍♂️  - deleting

File: packages/gatsby/package.json [hunk 4]
```diff
@@ -173,13 +174,16 @@
...
+    "build:cache-dir": "babel src/cache-dir --out-dir cache-dir",
     "build:internal-plugins": "copyfiles -u 1 src/internal-plugins/**/package.json dist",
     "build:rawfiles": "copyfiles -u 1 src/internal-plugins/**/raw_* dist",
-    "build:src": "babel src --out-dir dist --source-maps --ignore **/gatsby-cli.js,**/raw_*,**/__tests__",
+    "build:src": "babel src --out-dir dist --source-maps --ignore **/gatsby-cli.js,**/raw_*,**/__tests__,**/cache-dir",
     "clean-test-bundles": "find test/ -type f -name bundle.js* -exec rm -rf {} +",
     "prepublish": "cross-env NODE_ENV=production npm run build",
     "test-coverage": "node_modules/.bin/nyc --reporter=lcov --reporter=text npm test",
-    "watch": "rimraf dist && mkdir dist && npm run build:internal-plugins && npm run build:rawfiles && npm run build:src -- --watch"
+    "watch": "rimraf dist && rimraf cache-dir && mkdir dist && npm run build:internal-plugins && npm-run-all -p watch:*",
```
Comment (m-allanson) [inline packages/gatsby/package.json]: `npm run build:rawfiles ` is missing here. I'll add a commit.

File: packages/gatsby/src/cache-dir/static-entry.js (hunk 1 above)
Comment (m-allanson) [reply packages/gatsby/src/cache-dir/static-entry.js]: This looks ok to delete - the related code was removed in https://github.com/gatsbyjs/gatsby/commit/7983e271a7d27a029108508edd64b45c5466c17e#diff-917ba78a52f29b1a1fe42be34d81fc83

File: packages/gatsby/package.json (hunk 4 above)
Comment (pieh) [reply packages/gatsby/package.json]: ops, thanks for catching this!

**Comment by m-allanson** [comment]
Thanks @pieh 🎉

**Pull request merged**